Sming/Components/Hosted/include/Hosted/Client.h

```diff
@@ -0,0 +1,178 @@
+/****
+ * Sming Framework Project - Open Source framework for high efficiency native ESP8266 development.
+ * Created 2015 by Skurydin Alexey
+ * http://github.com/SmingHub/Sming
+ * All files of the Sming Core are provided under the LGPL v3 license.
+ *
+ * Client.h
+ *
+ * @author 2021 Slavey Karadzhov <slav@attachix.com>
+ *
+ *
+ ****/
+
+#pragma once
+
+#ifndef ARCH_HOST
+#error "Hosted::Client can be used only on the Host architecture!"
+#endif
+
+#include <Stream.h>
+#include <WHashMap.h>
+#include <WString.h>
+#include <simpleRPC.h>
+#include <simpleRPC/parser.h>
+#include <memory>
+#include <hostlib/hostlib.h>
+#include <hostlib/hostmsg.h>
+
+using namespace simpleRPC;
+
+namespace Hosted
+{
+constexpr int COMMAND_NOT_FOUND = -1;
+
+class Client
+{
+public:
+	using RemoteCommands = HashMap<String, uint8_t>;
+
+	Client(Stream& stream) : stream(stream)
+	{
+	}
+
+	/**
+	 * @brief Method to send commands to the remote server
+	 * @param functionName
+	 * @param variable arguments
+	 *
+	 * @retval true on success
+	 */
+	template <typename... Args> bool send(const char* functionName, Args... args)
+	{
+		uint8_t functionId = getFunctionId(functionName);
+
+		rpcPrint(stream, functionId, args...);
+
+		return true;
+	}
+
+	/**
+	 * @brief This method will block the execution until a message is detected
+	 * @retval HostedCommand
+	 */
+	template <typename R> R wait()
+	{
+		size_t neededBytes = sizeof(R);
+
+		while(stream.available() < int(neededBytes)) {
+			stream.flush();
+			host_main_loop();
+		}
+
+		std::shared_ptr<uint8_t> data(new uint8_t[neededBytes]);
```

Comment:
@mikee47 I would like to get a stream of bytes and cast them to the expected return type without also having a memory leak. So far I came up with this code here but is there a better way to do it in C++?

Reply:
How about this:

```
R result;
stream.readBytes(&result, sizeof(result));
return result;
```

Reply:
Wow! Cannot be that simple :)